akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala

```diff
@@ -212,10 +229,13 @@ private[http2] class Http2ServerDemux(http2Settings: Http2CommonSettings, initia
 
       setHandler(substreamIn, new InHandler {
         def onPush(): Unit = {
-          val sub = grab(substreamIn)
-          pull(substreamIn)
+          val sub: Http2SubStream = grab(substreamIn)
```

Comment:
The logic to pull must go somewhere else:

 * on the server side, a substream received here, will not open a new stream (but actually might in the future if we implement PUSH_PROMISE by letting responses "suggest" PUSH_PROMISES)
 * the stage has lots of entrypoints because of all the ports and substream handlers, etc. E.g. an action might be required if we read the last frame of a response in `IncomingStreamBuffer.onUpstreamFinish`.

I think we should add an action to the stream handling `updateState` which (on the client) tries to pull if there's capacity for outgoing capacity.

Reply:
(This also one reason we need the refactoring, because it is much more principled about going through the state machine at all times)